### PR TITLE
Add TCX to codoon_sync.py

### PR DESCRIPTION
**Pull request by lbp0**
add argument --with-tcx

### CODE REVIEW

**Comment by yihong0618** [comment]
Than is cool  thanks, @ben-29 can you help the review?

File: scripts/codoon_sync.py [hunk 1]
```diff
@@ -18,6 +18,15 @@
 
 from utils import adjust_time_to_utc
 
+import numpy as np
```
Comment (yihong0618) [inline scripts/codoon_sync.py:21]: seems we need to add numpy to requirements

File: scripts/codoon_sync.py [hunk 2]
```diff
@@ -75,6 +84,139 @@ def download_codoon_gpx(gpx_data, log_id):
...
 
+def autorm():   # auto remove data.db[for debugging only]
+    if os.path.exists("data.db"):
+        os.remove("data.db")
+    if os.path.exists("data.db-journal"):
+        os.remove("data.db-journal")
+        
+def set_array(fit_array,array_time,array_bpm,array_lati,array_longi):
+    fit_data = np.array((array_time,array_bpm,array_lati,array_longi),dtype = FitType)
+    if fit_array is None:
```
Comment (yihong0618) [inline scripts/codoon_sync.py:95]: fit_array can be None here?

**Review by yihong0618 — COMMENTED** [review]
Nice work. we will take test it later

**Comment by yihong0618** [comment]
please fix the lint ~

File: scripts/codoon_sync.py (hunk 2 above)
Comment (lbp0) [reply scripts/codoon_sync.py:95]: yep, in line 191, set None as default

File: scripts/codoon_sync.py [hunk 3]
```diff
@@ -75,6 +84,139 @@ def download_codoon_gpx(gpx_data, log_id):
         pass
 
 
+def autorm():   # auto remove data.db[for debugging only]
```
Comment (ben-29) [inline scripts/codoon_sync.py:87]: do we need this function anymore?

File: scripts/codoon_sync.py [hunk 4]
```diff
@@ -75,6 +84,139 @@ def download_codoon_gpx(gpx_data, log_id):
...
+    activity_creator_name = ET.Element("Name")
+    activity_creator_name.text = "咕咚"
+    activity_creator.append(activity_creator_name)
+    #   Lap
+    fit_start_time_CH = run_data["start_time"]
+        # parse to time array
+    time_array = time.strptime(fit_start_time_CH, "%Y-%m-%dT%H:%M:%S")
+        # parse to unix timestamp
+    unix_time = int(time.mktime(time_array))
+    unix_time = unix_time - 28800   # move to UTC[for UTC+08:00 only]
```
Comment (ben-29) [inline scripts/codoon_sync.py:139]: you can use `adjust_time_to_utc` function

File: scripts/codoon_sync.py [hunk 5]
```diff
@@ -75,6 +84,139 @@ def download_codoon_gpx(gpx_data, log_id):
...
+            lati = ET.Element("LatitudeDegrees")
+            lati.text = bytes.decode(i["lati"])
+            position.append(lati)
+        #   LongitudeDegrees
+            longi = ET.Element("LongitudeDegrees")
+            longi.text = bytes.decode(i["longi"])
+            position.append(longi)
+    
+    # write to TCX file
+    tree.write("../TCX_OUT/" + fit_id + ".tcx", encoding='utf-8', xml_declaration=True)
```
Comment (ben-29) [inline scripts/codoon_sync.py:186]: use `GPX_FOLDER` instead

**Review by ben-29 — CHANGES_REQUESTED** [review]
Good job. please check the comments and fix the lint with `black`, Great thanks.

File: scripts/codoon_sync.py (hunk 3 above)
Comment (lbp0) [reply scripts/codoon_sync.py:87]: ok i will remove this

File: scripts/codoon_sync.py (hunk 5 above)
Comment (lbp0) [reply scripts/codoon_sync.py:186]: do i need to export tcx files to gpx folder? and where is that constant please, i didn't find it

Comment (ben-29) [reply scripts/codoon_sync.py:186]: https://github.com/yihong0618/running_page/blob/3ae13b477ef75034ff455ae1fe3163f041afb610/scripts/config.py#L11
Sorry for my typo, it's `TCX_FOLDER`

Comment (lbp0) [reply scripts/codoon_sync.py:186]: got it thx

File: requirements.txt [hunk 6]
```diff
@@ -19,3 +19,5 @@ lxml
 eviltransform
 stravaweblib
 tenacity
+numpy
+tzlocal
```
Comment (yihong0618) [inline requirements.txt:23]: we not not need this, you can use other function in this repo, that can get local time

File: scripts/utils.py [hunk 7]
```diff
@@ -23,6 +23,28 @@ def adjust_time_to_utc(time, tz_name):
...
+    # once we decide to move on to python v3.7+
+    ts_fmts = ["%Y-%m-%dT%H:%M:%S", "%Y-%m-%dT%H:%M:%S.%f"]
+
+    for ts_fmt in ts_fmts:
+        try:
+            # performance with using exceptions
+            # shouldn't be an issue since it's an offline cmdline tool
+            return datetime.strptime(ts, ts_fmt)
+        except ValueError:
+            pass
```
Comment (yihong0618) [inline scripts/utils.py:43]: maybe print something?

**Review by yihong0618 — APPROVED** [review]
LGTM

**Comment by yihong0618** [comment]
@ben-29 most LGTM
@lbp0 Thank you very much, and welcome to open source world~

**Review by ben-29 — COMMENTED** [review]
I have done a little bit of testing, but the tcx files fail to save. 
After changing a few lines of code. It did save tcx files successfully. Well done and do the final change~
You can use my auth-token for testing: 2ec719f753af36a413e2158f82d7d742 fef1a5e9-daac-4af4-bac5-b78d887a4a08

File: scripts/codoon_sync.py [hunk 8]
```diff
@@ -75,6 +100,161 @@ def download_codoon_gpx(gpx_data, log_id):
...
+        time_stamp = utc.strftime("%Y-%m-%dT%H:%M:%SZ")
+        # to time array
+        time_array = time.strptime(time_stamp, "%Y-%m-%dT%H:%M:%SZ")
+        # to unix timestamp
+        unix_time = int(time.mktime(time_array))
+        # set GPS data
+        # if the track point which has the same time has been added
+        for i in fit_array:
+            if i["time"] == unix_time:
+                print("1")
```
[the remark this answers is not in the document]
Comment (lbp0) [reply scripts/codoon_sync.py:244]: sorry i didn't notice this lol

**Comment by lbp0** [comment]
> @ben-29 most LGTM
> @lbp0 Thank you very much, and welcome to open source world~

thanks so much,both of you

File: scripts/codoon_sync.py [hunk 9]
```diff
@@ -75,6 +100,161 @@ def download_codoon_gpx(gpx_data, log_id):
...
+    for single_time, single_bpm in own_heart_rate.items():
+        single_time = adjust_timestemp_to_utc(single_time, str(get_localzone()))
+        # set bpm data
+        fit_array = set_array(fit_array, single_time, single_bpm, None, None)
+    # get single track point
+    for point in own_points:
+        repeat_flag = False
+        time_stamp = point.get("time_stamp")
+        latitude = point.get("latitude")
+        longitude = point.get("longitude")
```
[the remark this answers is not in the document]
Comment (lbp0) [reply scripts/codoon_sync.py:231]: I don't decide to add elevations for this time, just step by step. I will do it after finish this merging :P

File: requirements.txt (hunk 6 above)
Comment (lbp0) [reply requirements.txt:23]: no i use it to get ueser's timezone. or we have a function to get this in this repo?

File: scripts/codoon_sync.py [hunk 10]
```diff
@@ -75,6 +100,161 @@ def download_codoon_gpx(gpx_data, log_id):
         pass
 
 
+def set_array(fit_array, array_time, array_bpm, array_lati, array_longi):
+    fit_data = np.array((array_time, array_bpm, array_lati, array_longi), dtype=FitType)
+    if fit_array is None:
+        fit_array = fit_data
```
[the remark this answers is not in the document]
Comment (lbp0) [reply scripts/codoon_sync.py:106]: it works good in my environment so i didn't learn more about this. thank you so much for your advice

File: scripts/codoon_sync.py [hunk 11]
```diff
@@ -75,6 +100,161 @@ def download_codoon_gpx(gpx_data, log_id):
...
+        TCX_FOLDER + "/" + fit_id + ".tcx", encoding="utf-8", xml_declaration=True
+    )
+
+
+def tcx_job(run_data):
+    # fit struct array
+    fit_array = None
+
+    # raw data
+    own_heart_rate = run_data["heart_rate"]  # bpm key-value
```
[the remark this answers is not in the document]
Comment (ben-29) [reply scripts/codoon_sync.py:219]: Try run with `--with-tcx --from-auth-token  2ec719f753af36a413e2158f82d7d742 fef1a5e9-daac-4af4-bac5-b78d887a4a08 `, you will see what's wrong.
Do the final kick and we will merge when the tcx saves as expected.

**Comment by lbp0** [comment]
> I have done a little bit of testing, but the tcx files fail to save. After changing a few lines of code. It did save tcx files successfully. Well done and do the final change~ You can use my auth-token for testing: 2ec719f753af36a413e2158f82d7d742 fef1a5e9-daac-4af4-bac5-b78d887a4a08

now all of your runs can be exported successfully. thanks for your token!!!

**Review by ben-29 — APPROVED** [review]
Well done. Thanks for contributing.

**Comment by yihong0618** [comment]
@lbp0 Thanks a lot, its amazing work, we will use your work that add `--with-tcx` to other `sync` either.~

**Pull request merged**